### PR TITLE
Add Maven resolving support to the Bindgen tool

### LINKED ISSUE #23066: [Bindgen Tool] Integrate Maven support for the tool
**Description:**
The Bindgen tool should be able to identify whether the tool is being executed inside a project and if so, it should be able to use Maven support and obtain dependencies. The tool should also be able to update the Ballerina.toml file with the Maven dependencies.

**Steps to reproduce:**

**Affected Versions:**

**OS, DB, other environment details and versions:**

**Related Issues (optional):**

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

### PR DESCRIPTION
**Pull request by IrushiL**
## Purpose
This PR adds the following functionality:
- Ability to identify if the tool is being run inside a Ballerina project or if the output path parameter points to a location inside a Ballerina project.
- Based on the above knowledge, the `Ballerina.toml` file is read in order to identify jars and maven dependencies specified under `platform.libraries`. These are then added to the classpath of the `Bindgen` tool.
- Improves the error message of failed classpath additions by printing the jar/directory through which the classpath addition is attempted.
- Adds the Maven support to the `Bindgen` tool. If the tool detects a Ballerina project, the Maven dependencies will get resolved into `target/maven-repo` directory. Else if an output parameter is provided, the dependencies will get resolved into a directory created under `target/maven-repo`. If the output parameter is also not provided the `target/maven-repo` directory will be created inside the directory where the tool is being executed.
- The tool also updates the `Ballerina.toml` file with the Maven dependency provided through the `Bindgen` command and the transitive dependencies resolved through it.

Fixes #23066

## Approach
Maven support is added as follows:
```ballerina
ballerina bindgen [(-cp|--classpath) <classpath>...]
                  [(-mvn|--maven) <groupId>:<artifactId>:<version>]
                  [(-o|--output) <output>]
                  (<class-name>...)
```

## Check List 
- [X] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: misc/ballerina-bindgen/src/main/java/org/ballerinalang/bindgen/command/BindgenCommand.java [hunk 1]
```diff
@@ -38,6 +42,8 @@
 
     private static final PrintStream outStream = System.out;
     private static final PrintStream outError = System.err;
+    private Path sourceRootPath = Paths.get(System.getProperty("user.dir"));
```
Comment (pramodya1994) [inline misc/ballerina-bindgen/src/main/java/org/ballerinalang/bindgen/command/BindgenCommand.java:45]:
```suggestion
    private Path sourceRootPath = Paths.get(System.getProperty(USER_DIR));
```
Can we use `USER_DIR` from `BindgenConstants`?

File: misc/ballerina-bindgen/src/main/java/org/ballerinalang/bindgen/command/BindgenCommand.java [hunk 2]
```diff
@@ -38,6 +42,8 @@
...
+    private Path targetOutputPath = sourceRootPath;
```
Comment (pramodya1994) [inline misc/ballerina-bindgen/src/main/java/org/ballerinalang/bindgen/command/BindgenCommand.java:46]: Is there a possibility to use one variable for `targetOutputPath` and `sourceRootPath`?

File: misc/ballerina-bindgen/src/main/java/org/ballerinalang/bindgen/command/BindgenCommand.java [hunk 3]
```diff
@@ -87,6 +114,20 @@ public void execute() {
...
+                outError.println("\nError in the maven dependency provided.\n");
+                outStream.println(BINDGEN_CMD);
+                outStream.println("\nUse 'ballerina bindgen --help' for more information on the command.");
+                return;
+            }
+            bindingsGenerator.setMvnGroupId(mvnDependency[0]);
+            bindingsGenerator.setMvnArtifactId(mvnDependency[1]);
+            bindingsGenerator.setMvnVersion(mvnDependency[2]);
+        }
+
```
Comment (pramodya1994) [inline misc/ballerina-bindgen/src/main/java/org/ballerinalang/bindgen/command/BindgenCommand.java:130]: Shall we throw `UnsupportedOperationException` in `setParentCmdParser(...)` in line number 164?

File: misc/ballerina-bindgen/src/main/java/org/ballerinalang/bindgen/command/BindingsGenerator.java [hunk 4]
```diff
@@ -79,6 +88,27 @@
     private static Map<String, String> failedClassGens = new HashMap<>();
 
     void generateJavaBindings() throws BindgenException {
+        if (projectRoot != null) {
+            Manifest manifest = TomlParserUtils.getManifest(projectRoot);
+            if (manifest != null) {
+                List<Library> platformLibraries = manifest.getPlatform().getLibraries();
+                if (platformLibraries != null) {
+                    for (Library library: platformLibraries) {
```
Comment (keizer619) [inline misc/ballerina-bindgen/src/main/java/org/ballerinalang/bindgen/command/BindingsGenerator.java:96]: @IrushiL There has to be a space before colon. Usually these should be detected by check style validations.

File: misc/ballerina-bindgen/src/main/java/org/ballerinalang/bindgen/command/BindingsGenerator.java [hunk 5]
```diff
@@ -79,6 +88,27 @@
...
+                        if (library.path != null) {
+                            classPaths.add(Paths.get(projectRoot.toString(), library.path).toString());
+                        } else if (library.groupId != null && library.artifactId != null && library.version != null) {
+                            mavenResolver(library.groupId, library.artifactId, library.version, projectRoot);
+                        }
+                    }
+                }
+            }
+        }
```
Comment (pramodya1994) [inline misc/ballerina-bindgen/src/main/java/org/ballerinalang/bindgen/command/BindingsGenerator.java:105]: Isn't it better to move this logic to a separate method? So we can reduce the complexity of the `generateJavaBindings` method.

Comment (IrushiL) [reply misc/ballerina-bindgen/src/main/java/org/ballerinalang/bindgen/command/BindingsGenerator.java]: Done

File: misc/ballerina-bindgen/src/main/java/org/ballerinalang/bindgen/command/BindingsGenerator.java (hunk 4 above)
Comment (IrushiL) [reply misc/ballerina-bindgen/src/main/java/org/ballerinalang/bindgen/command/BindingsGenerator.java]: Fixed it.

File: misc/ballerina-bindgen/src/main/java/org/ballerinalang/bindgen/command/BindgenCommand.java (hunk 3 above)
Comment (IrushiL) [reply misc/ballerina-bindgen/src/main/java/org/ballerinalang/bindgen/command/BindgenCommand.java:130]: All the other commands seem to be following this norm. I think we can keep this as it is.

File: misc/ballerina-bindgen/src/main/java/org/ballerinalang/bindgen/command/BindgenCommand.java (hunk 2 above)
Comment (IrushiL) [reply misc/ballerina-bindgen/src/main/java/org/ballerinalang/bindgen/command/BindgenCommand.java]: Done.

File: misc/ballerina-bindgen/src/main/java/org/ballerinalang/bindgen/command/BindgenCommand.java (hunk 1 above)
Comment (IrushiL) [reply misc/ballerina-bindgen/src/main/java/org/ballerinalang/bindgen/command/BindgenCommand.java]: Yes, fixed.

**Comment by IrushiL** [comment]
> I added some comments

**Comment by IrushiL** [comment]
> I added some comments

Fixed

**Pull request merged**